File: sdk/acp_client/policy.py
"""Policy-as-code surface.

The ACP policy file is a YAML document checked into the customer's repo. It
defines what each agent is allowed to do. The CLI validates it before deploy;
the gateway enforces it at runtime.

Schema (v1):

    version: 1
    agent: <agent-id>
    allow:
      - tool: <tool-name>          # required
        when:                      # optional predicate
          payload.<field>: <regex>
    deny:
      - tool: <tool-name>
    autonomy:
      max_actions_per_minute: <int>
      max_blast_radius: <int>
      require_approval_for: [<tool-name>, ...]
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .errors import PolicyError

SUPPORTED_VERSIONS = {1}
VALID_TOP_LEVEL = {"version", "agent", "allow", "deny", "autonomy"}


@dataclass
class Rule:
    tool: str
    when: dict[str, str] = field(default_factory=dict)  # field → regex


@dataclass
class Autonomy:
    max_actions_per_minute: int | None = None
    max_blast_radius: int | None = None
    require_approval_for: list[str] = field(default_factory=list)


@dataclass
class Policy:
    version: int
    agent: str
    allow: list[Rule]
    deny: list[Rule]
    autonomy: Autonomy

    def matches_allow(self, tool: str, payload: dict[str, Any]) -> bool:
        return any(_rule_matches(r, tool, payload) for r in self.allow)

    def matches_deny(self, tool: str, payload: dict[str, Any]) -> bool:
        return any(_rule_matches(r, tool, payload) for r in self.deny)


def load_policy(path: str | Path) -> Policy:
    """Parse + validate a policy file. Raises PolicyError on any issue."""
    import yaml  # local import: only required when actually loading a file

    p = Path(path)
    if not p.exists():
        raise PolicyError(f"policy file not found: {p}")
    try:
        raw = yaml.safe_load(p.read_text())
    except yaml.YAMLError as e:
        raise PolicyError(f"invalid YAML in {p}: {e}") from e
    return validate_policy(raw)


def validate_policy(raw: Any) -> Policy:
    """Validate a parsed dict and return a Policy. Raises PolicyError."""
    if not isinstance(raw, dict):
        raise PolicyError("policy root must be a mapping")

    unknown = set(raw) - VALID_TOP_LEVEL
    if unknown:
        raise PolicyError(f"unknown top-level keys: {sorted(unknown)}")

    version = raw.get("version")
    if version not in SUPPORTED_VERSIONS:
        raise PolicyError(f"unsupported policy version: {version!r}")

    agent = raw.get("agent")
    if not isinstance(agent, str) or not agent:
        raise PolicyError("`agent` must be a non-empty string")

    allow = [_parse_rule(r, "allow", i) for i, r in enumerate(raw.get("allow") or [])]
    deny = [_parse_rule(r, "deny", i) for i, r in enumerate(raw.get("deny") or [])]

    autonomy_raw = raw.get("autonomy") or {}
    if not isinstance(autonomy_raw, dict):
        raise PolicyError("`autonomy` must be a mapping")
    autonomy = Autonomy(
        max_actions_per_minute=autonomy_raw.get("max_actions_per_minute"),
        max_blast_radius=autonomy_raw.get("max_blast_radius"),
        require_approval_for=list(autonomy_raw.get("require_approval_for") or []),
    )

    return Policy(version=version, agent=agent, allow=allow, deny=deny, autonomy=autonomy)


def _parse_rule(raw: Any, kind: str, idx: int) -> Rule:
    if not isinstance(raw, dict):
        raise PolicyError(f"{kind}[{idx}] must be a mapping")
    tool = raw.get("tool")
    if not isinstance(tool, str) or not tool:
        raise PolicyError(f"{kind}[{idx}].tool must be a non-empty string")
    when = raw.get("when") or {}
    if not isinstance(when, dict):
        raise PolicyError(f"{kind}[{idx}].when must be a mapping")
    # Compile each regex eagerly so syntax errors surface at validate time.
    for field_path, pattern in when.items():
        if not isinstance(pattern, str):
            raise PolicyError(f"{kind}[{idx}].when.{field_path} must be a string regex")
        try:
            re.compile(pattern)
        except re.error as e:
            raise PolicyError(f"{kind}[{idx}].when.{field_path}: invalid regex: {e}") from e
    return Rule(tool=tool, when=when)


def _rule_matches(rule: Rule, tool: str, payload: dict[str, Any]) -> bool:
    if rule.tool != tool and rule.tool != "*":
        return False
    for field_path, pattern in rule.when.items():
        value = _resolve(payload, field_path)
        if value is None or not re.search(pattern, str(value)):
            return False
    return True


def _resolve(obj: Any, path: str) -> Any:
    cur: Any = obj
    for part in path.split("."):
        if part == "payload":
            continue
        if isinstance(cur, dict):
            cur = cur.get(part)
        else:
            return None
    return cur

File: sdk/acp-js/src/policy.ts
import { readFileSync } from "node:fs";
import { parse as parseYaml } from "yaml";
import { PolicyError } from "./errors.js";

export interface Rule {
  tool: string;
  when: Record<string, string>;
}

export interface Autonomy {
  maxActionsPerMinute?: number;
  maxBlastRadius?: number;
  requireApprovalFor: string[];
}

export interface Policy {
  version: number;
  agent: string;
  allow: Rule[];
  deny: Rule[];
  autonomy: Autonomy;
}

const SUPPORTED_VERSIONS = new Set<number>([1]);
const VALID_TOP_LEVEL = new Set<string>(["version", "agent", "allow", "deny", "autonomy"]);

export function loadPolicy(path: string): Policy {
  let raw: unknown;
  try {
    raw = parseYaml(readFileSync(path, "utf-8"));
  } catch (e) {
    throw new PolicyError(`invalid YAML in ${path}: ${(e as Error).message}`);
  }
  return validatePolicy(raw);
}

export function validatePolicy(raw: unknown): Policy {
  if (!isObject(raw)) throw new PolicyError("policy root must be a mapping");

  const unknown = Object.keys(raw).filter((k) => !VALID_TOP_LEVEL.has(k));
  if (unknown.length > 0) {
    throw new PolicyError(`unknown top-level keys: ${JSON.stringify(unknown.sort())}`);
  }

  const version = raw["version"];
  if (typeof version !== "number" || !SUPPORTED_VERSIONS.has(version)) {
    throw new PolicyError(`unsupported policy version: ${JSON.stringify(version)}`);
  }

  const agent = raw["agent"];
  if (typeof agent !== "string" || agent.length === 0) {
    throw new PolicyError("`agent` must be a non-empty string");
  }

  const allow = parseRuleList(raw["allow"], "allow");
  const deny = parseRuleList(raw["deny"], "deny");

  const autonomyRaw = raw["autonomy"];
  if (autonomyRaw !== undefined && !isObject(autonomyRaw)) {
    throw new PolicyError("`autonomy` must be a mapping");
  }
  const ar = (autonomyRaw ?? {}) as Record<string, unknown>;
  const autonomy: Autonomy = {
    maxActionsPerMinute: typeof ar["max_actions_per_minute"] === "number" ? (ar["max_actions_per_minute"] as number) : undefined,
    maxBlastRadius: typeof ar["max_blast_radius"] === "number" ? (ar["max_blast_radius"] as number) : undefined,
    requireApprovalFor: Array.isArray(ar["require_approval_for"]) ? (ar["require_approval_for"] as string[]) : [],
  };

  return { version, agent, allow, deny, autonomy };
}

function parseRuleList(raw: unknown, kind: string): Rule[] {
  if (raw === undefined || raw === null) return [];
  if (!Array.isArray(raw)) throw new PolicyError(`\`${kind}\` must be a list`);
  return raw.map((r, i) => parseRule(r, kind, i));
}

function parseRule(raw: unknown, kind: string, idx: number): Rule {
  if (!isObject(raw)) throw new PolicyError(`${kind}[${idx}] must be a mapping`);
  const tool = raw["tool"];
  if (typeof tool !== "string" || tool.length === 0) {
    throw new PolicyError(`${kind}[${idx}].tool must be a non-empty string`);
  }
  const whenRaw = raw["when"];
  if (whenRaw !== undefined && !isObject(whenRaw)) {
    throw new PolicyError(`${kind}[${idx}].when must be a mapping`);
  }
  const when: Record<string, string> = {};
  for (const [k, v] of Object.entries(whenRaw ?? {})) {
    if (typeof v !== "string") {
      throw new PolicyError(`${kind}[${idx}].when.${k} must be a string regex`);
    }
    try {
      new RegExp(v);
    } catch (e) {
      throw new PolicyError(`${kind}[${idx}].when.${k}: invalid regex: ${(e as Error).message}`);
    }
    when[k] = v;
  }
  return { tool, when };
}

function isObject(x: unknown): x is Record<string, unknown> {
  return typeof x === "object" && x !== null && !Array.isArray(x);
}
